### PR TITLE
Fix the bug that maybe causing memory leak and repeated traceId when use gateway-2.1.x-plugin

### PR DESCRIPTION
**Pull request by xiarixigua**
### Fix the bug that maybe causing memory leak and repeated traceId when use gateway-2.1.x-plugin
- [x] Add a unit test to verify that the fix works.
- [x] Explain briefly why the bug exists and how to fix it.
use custom filter like this in spring cloud gateway 2.2.3.RELEASE, cause the bug. 
```
@Component("customFilter")
public class CustomFilter implements GlobalFilter, Ordered {

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        return chain.filter(exchange).then(Mono.defer(() -> {
            // do something
            return chain.filter(exchange);
        }));
    }

    @Override
    public int getOrder() {
        return NettyWriteResponseFilter.WRITE_RESPONSE_FILTER_ORDER;
    }
}
```

because the custom filter causing NettyRoutingFilterInterceptor.beforeMethod called twice, but HttpClientFinalizerSendInterceptor.beforeMethod just called once in request. So TracingContext.activeSpanStack size not be zero when request trace end.

Fix it: do nothing when second call NettyRoutingFilterInterceptor.beforeMethod:

```
    private static final String NETTY_ROUTING_FILTERED_ATTR = NettyRoutingFilterInterceptor.class.getName() + ".alreadyFiltered";

    @Override
    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
                             MethodInterceptResult result) throws Throwable {
        ServerWebExchange exchange = (ServerWebExchange) allArguments[0];
        if (isAlreadyFiltered(exchange)) {
            return;
        }

        setAlreadyFiltered(exchange);

        EnhancedInstance enhancedInstance = getInstance(exchange);

        AbstractSpan span = ContextManager.createLocalSpan("SpringCloudGateway/RoutingFilter");
        if (enhancedInstance != null && enhancedInstance.getSkyWalkingDynamicField() != null) {
            ContextManager.continued((ContextSnapshot) enhancedInstance.getSkyWalkingDynamicField());
        }
        span.setComponent(SPRING_CLOUD_GATEWAY);
    }

    private static void setAlreadyFiltered(ServerWebExchange exchange) {
        exchange.getAttributes().put(NETTY_ROUTING_FILTERED_ATTR, true);
    }

    private static boolean isAlreadyFiltered(ServerWebExchange exchange) {
        return exchange.getAttributeOrDefault(NETTY_ROUTING_FILTERED_ATTR, false);
    }
```

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
Thanks for reporting and sharing the context. Could you do the following to make this PR ready?
1. Update changes.md in the root
2. Find several existing issues, I remember there are several related to gateway plugins, but I am not sure whether this PR could fix them.

One question, when does RPC happens, after 1st or 2nd time `NettyRoutingFilterInterceptor` called?

**Comment by wu-sheng** [comment]
`NettyRoutingFilterInterceptor` exists in 2.0, 2.1, 3.0 plugins, could you verify them all? Because they seem in very similar way.

**Comment by xiarixigua** [comment]
> `NettyRoutingFilterInterceptor` exists in 2.0, 2.1, 3.0 plugins, could you verify them all? Because they seem in very similar way.

2.0-plugin can not cause this bug.
2.1-plugin and 3.x-plugin can cause this bug.

**Comment by wu-sheng** [comment]
> > `NettyRoutingFilterInterceptor` exists in 2.0, 2.1, 3.0 plugins, could you verify them all? Because they seem in very similar way.
> 
> 2.0-plugin can not cause this bug. 2.1-plugin and 3.x-plugin can cause this bug.

Thanks for the feedback. Please fix them as well, thanks.

**Comment by xiarixigua** [comment]
> Thanks for reporting and sharing the context. Could you do the following to make this PR ready?
> 
> 1. Update changes.md in the root
> 2. Find several existing issues, I remember there are several related to gateway plugins, but I am not sure whether this PR could fix them.
> 
> One question, when does RPC happens, after 1st or 2nd time `NettyRoutingFilterInterceptor` called?

1. CHANGES.md updated
2. i'm not find other issues when use gateway plugins

RPC happens after 1st time `NettyRoutingFilterInterceptor` called

**Comment by wu-sheng** [comment]
About (2), I think SkyWalking GitHub issues list.

**Comment by wu-sheng** [comment]
```
Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.503 s <<< FAILURE! - in org.apache.skywalking.apm.plugin.spring.cloud.gateway.v21x.NettyRoutingFilterInterceptorTest
Error:  testAlreadyFiltered(org.apache.skywalking.apm.plugin.spring.cloud.gateway.v21x.NettyRoutingFilterInterceptorTest)  Time elapsed: 0.84 s  <<< ERROR!
java.lang.NullPointerException
	at org.apache.skywalking.apm.plugin.spring.cloud.gateway.v21x.NettyRoutingFilterInterceptorTest.testAlreadyFiltered(NettyRoutingFilterInterceptorTest.java:137)
```

It looks like you missed the UT.

**Comment by wu-sheng** [comment]
This plans to fix https://github.com/apache/skywalking/issues/7923 https://github.com/apache/skywalking/issues/7922 https://github.com/apache/skywalking/issues/7658

File: apm-sniffer/optional-plugins/optional-spring-plugins/optional-spring-cloud/gateway-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/cloud/gateway/v3x/NettyRoutingFilterInterceptor.java [hunk 1]
```diff
@@ -51,6 +59,14 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
         span.setComponent(SPRING_CLOUD_GATEWAY);
     }
 
+    private static void setAlreadyFiltered(ServerWebExchange exchange) {
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/optional-spring-plugins/optional-spring-cloud/gateway-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/cloud/gateway/v3x/NettyRoutingFilterInterceptor.java:62]:
```suggestion
    private static void isTraced(ServerWebExchange exchange) {
```

File: apm-sniffer/optional-plugins/optional-spring-plugins/optional-spring-cloud/gateway-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/cloud/gateway/v3x/NettyRoutingFilterInterceptor.java [hunk 2]
```diff
@@ -51,6 +59,14 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
...
+        exchange.getAttributes().put(NETTY_ROUTING_FILTERED_ATTR, true);
+    }
+
+    private static boolean isAlreadyFiltered(ServerWebExchange exchange) {
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/optional-spring-plugins/optional-spring-cloud/gateway-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/cloud/gateway/v3x/NettyRoutingFilterInterceptor.java:66]:
```suggestion
    private static boolean isTraced(ServerWebExchange exchange) {
```

**Comment by wu-sheng** [comment]
You can't just apply my changes, this is a change tip, but I didn't change all relative codes.

**Comment by xiarixigua** [comment]
> You can't just apply my changes, this is a change tip, but I didn't change all relative codes.

ok, i will finish other relative codes.

**Comment by xiarixigua** [comment]
> Have you fixed the broken UT?

fixed

**Comment by lkxiaolou** [comment]
> This plans to fix [apache/skywalking#7923](https://github.com/apache/skywalking/issues/7923) [apache/skywalking#7922](https://github.com/apache/skywalking/issues/7922) [apache/skywalking#7658](https://github.com/apache/skywalking/issues/7658)
> 
> @lkxiaolou I mean this ☝️

Seems to miss me 😀

**Comment by xiarixigua** [comment]
> Your UTs are still failing, please run and make it passed locally first.
> 
> https://github.com/apache/skywalking-java/runs/5754418883?check_suite_focus=true

`./mvnw clean package -Pall ` run successed this time

**Review by wu-sheng — APPROVED** [review]

**Comment by KAKALUOTEAAA** [comment]
> ### Fix the bug that maybe causing memory leak and repeated traceId when use gateway-2.1.x-plugin
> * [x]  Add a unit test to verify that the fix works.
> * [x]  Explain briefly why the bug exists and how to fix it.
>   use custom filter like this in spring cloud gateway 2.2.3.RELEASE, cause the bug.
> 
> ```
> @Component("customFilter")
> public class CustomFilter implements GlobalFilter, Ordered {
> 
>     @Override
>     public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
>         return chain.filter(exchange).then(Mono.defer(() -> {
>             // do something
>             return chain.filter(exchange);
>         }));
>     }
> 
>     @Override
>     public int getOrder() {
>         return NettyWriteResponseFilter.WRITE_RESPONSE_FILTER_ORDER;
>     }
> }
> ```
> 
> because the custom filter causing NettyRoutingFilterInterceptor.beforeMethod called twice, but HttpClientFinalizerSendInterceptor.beforeMethod just called once in request. So TracingContext.activeSpanStack size not be zero when request trace end.
> 
> Fix it: do nothing when second call NettyRoutingFilterInterceptor.beforeMethod:
> 
> ```
>     private static final String NETTY_ROUTING_FILTERED_ATTR = NettyRoutingFilterInterceptor.class.getName() + ".alreadyFiltered";
> 
>     @Override
>     public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
>                              MethodInterceptResult result) throws Throwable {
>         ServerWebExchange exchange = (ServerWebExchange) allArguments[0];
>         if (isAlreadyFiltered(exchange)) {
>             return;
>         }
> 
>         setAlreadyFiltered(exchange);
> 
>         EnhancedInstance enhancedInstance = getInstance(exchange);
> 
>         AbstractSpan span = ContextManager.createLocalSpan("SpringCloudGateway/RoutingFilter");
>         if (enhancedInstance != null && enhancedInstance.getSkyWalkingDynamicField() != null) {
>             ContextManager.continued((ContextSnapshot) enhancedInstance.getSkyWalkingDynamicField());
>         }
>         span.setComponent(SPRING_CLOUD_GATEWAY);
>     }
> 
>     private static void setAlreadyFiltered(ServerWebExchange exchange) {
>         exchange.getAttributes().put(NETTY_ROUTING_FILTERED_ATTR, true);
>     }
> 
>     private static boolean isAlreadyFiltered(ServerWebExchange exchange) {
>         return exchange.getAttributeOrDefault(NETTY_ROUTING_FILTERED_ATTR, false);
>     }
> ```
> 
> * [ ]  If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #.
> * [ ]  Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

"The custom filter causing NettyRoutingFilterInterceptor.beforeMethod called twice". I think this problem is caused by the incorrect use of filter syntax.
Is it correct to write "return chain.filter(exchange)" in "return chain.filter(exchange).then(Mono.defer(() -> {})"? 
`
@Component("customFilter")
public class CustomFilter implements GlobalFilter, Ordered {

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        return chain.filter(exchange).then(Mono.defer(() -> {
            // do something
            return chain.filter(exchange);
        }));
    }

    @Override
    public int getOrder() {
        return NettyWriteResponseFilter.WRITE_RESPONSE_FILTER_ORDER;
    }
}`
I think It could be "return Mono.empty();"
`
 @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        return chain.filter(exchange).then(Mono.defer(() -> {
            // do something
            return Mono.empty();
        }));
    }
`

**Comment by 2152865832** [comment]
When a user defines a custom filter that retries the request upon failure, both NettyRoutingFilterInterceptor.beforeMethod and HttpClientFinalizerSendInterceptor.beforeMethod are executed again during the retry. How should this situation be handled?

**Pull request merged**